### DISCUSSION TITLE
Win2D CanvasControl

### DISCUSSION THREAD

**Discussion by onslauth** [microsoft/WindowsAppSDK#2487] [Q&A]
Hi,

I am trying to use Win2D CanvasControl in a WinUI3 / AppSDK program. I have followed the instructions at: https://microsoft.github.io/Win2D/WinUI3/html/Introduction.htm

However, when I add the following code:
```
<Grid>
    <canvas:CanvasControl Draw="CanvasControl_Draw" ClearColor="CornflowerBlue"/>
</Grid>
```

It returns the error:
`Error	XLS0503	A value of type 'CanvasControl' cannot be added to a collection or dictionary of type 'UIElementCollection'.`

I have tried adding the control directly to `MainWindow`, `Page` and a `UserControl`, with and without the `Grid` control. I have also tried to place it inside a `Canvas` control, also to no avail.

Does anyone perhaps know of a solution?

**Comment by castorix** [comment]
> Does anyone perhaps know of a solution?

Not an answer, but I use Direct2D directly (in C#, VB, C++) instead of Win2D and it works fine with [CompositionTarget.Rendering](https://docs.microsoft.com/en-us/windows/winui/api/microsoft.ui.xaml.media.compositiontarget.rendering?view=winui-3.0)
Even without SwapChainPanel (actually bugged on resizing for scale > 100%), by replacing CreateSwapChainForComposition with **CreateSwapChainForHwnd**.
I could post a sampIe I made with Sprites animation (C#) if you don't find a solution...

![WinUI3_Direct2D_SpriteBatch](https://user-images.githubusercontent.com/22345506/167647913-d579b2dd-2377-4b5c-a57f-c251d07a399a.jpg)

**Comment by ChewWorker** [comment]
@onslauth can you verify which version of Win2D you are using? .uwp or .winui3?
